### PR TITLE
[inductor][cpp] epilogue support for gemm template

### PR DESCRIPTION
**Pull request by jgong5**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #126019

As part of #125683, this PR adds the epilogue support for c++ gemm template by reusing the c++ vector codegen on sub-slices of tensors. This is implemented by retracing the epilogue IR nodes with new ranges and offsets. The new `codegen_loop_bodies` and `codegen_functions` methods are added to c++ vector codegen for this purpose. This is leveraged by the `store_output` method of the template kernel for epilogue codegen and store to the final result.

cc @voznesenskym @penguinwu @EikanWang @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang

### CODE REVIEW

**Review by jansel — APPROVED** [review]

File: torch/_inductor/codegen/cpp_template_kernel.py [hunk 1]
```diff
@@ -61,8 +76,15 @@ def def_kernel(
...
         for sizevar in sizevars:
             self.args.sizevars[sizevar] = f"k{sizevar}"
-        cpp_argdefs, _, _ = self.args.cpp_argdefs()
-        return f"void {self.kernel_name}({', '.join(cpp_argdefs)})"
+
+        def hook():
+            cpp_argdefs, _, _ = self.args.cpp_argdefs()
+            return f"void {self.kernel_name}({', '.join(cpp_argdefs)})"
+
+        placeholder = "<DEFINE_KERNEL>"
```
[the remark this answers is not in the document]
Comment (jgong5) [reply torch/_inductor/codegen/cpp_template_kernel.py]: Thanks for the reminder. Fixed.

**Pull request closed without merging**